api/xds_protocol.rst

```diff
@@ -220,6 +220,61 @@ the ADS server, which will be used for all resources. The :ref:`ConfigSource
 The xDS transport Protocol
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
 
+.. _xds_protocol_resource_names:
+
+Resource names
+^^^^^^^^^^^^^^
+
+The xDS protocol is designed to provide a transport mechanism for opaque resources. Every resource
+has a name. This name may be opaque in simple or legacy configurations, the control plane may
+determine it and set this value in delivered configuration, e.g. RDS resource `Foo` may be a
+dependency of LDS resource `Bar`.
+
+For more advanced use cases, xDS has the concept of resource namespaces. With namespaces, resource
+names take the form of `<namespace>/<resource name>`. In this scheme, the `resource name` remains
+opaque, while the `namespace` is expected to have the form `<authority>/<qualifier>`.
+
+The `authority` field is distinguished. It is expected to be reverse DNS of the authority originating
+the resource. The rationale for the use of reverse DNS is to allow prefix namespace subscriptions
+with the ability of more specific sub-authorities to be delegated override capability. In future
+xDS extensions, the ability to sign resources may be added, this authority field will be the basis
+for validation.
+
+The `qualifier` field is opaque and may be used to further distinguish the resource namespace by
+the control plane.
+
+Resource namespaces are intended to support a number of use cases:
+
+* Delegating authority over some particular part of the global namespace, e.g. with CDS to allow
+  server X to serve up the `com.acme.foo` hierarchy.
+
+* Partitioning resource types into different classes, e.g. for distinct tenants. E.g. LDS
+  can use the namespace to determine which subset of the universe of known listeners to
+  deliver.
+
+* Support split-horizon naming of resources. Allow `com.acme.foo.bar` to override `com.acme.foo`
+  for a given resource namespace.
+
+* Avoiding the dependence on :ref:`per-node <envoy_api_msg_Core.Node>` identifiers and metadata
```

Comment:
Yes!  I would love it if we could move in the direction of trying to make wildcard query semantics go away.

Reply:
I think it really depends on the way that node is used. In some sense, node is really no different from the namespace depending on how a deployment uses it. I might slightly change the language here from "avoiding" to "provide an alternative" and be a bit more clear as to why we need both options.

Reply:
I'm not sure I agree completely here. Node can (and has) been used like a namespace, but it's a conflation that we should avoid in a clean design. Node describes the node, which is a global property of the client. This should really be invariant between different xDS requests (in particular on a shared ADS stream). Different resource types and resource fetches may have distinct namespaces. Hence the wish to re-express namespaces outside of Node.

Reply:
> Different resource types and resource fetches may have distinct namespaces.

The may, but in many (most?) deployments they will not. We should be very careful to not force extra complexity on folks that already find all of this incredibly complicated. Keying on node for all xDS will work for a lot of people.

Reply:
From the few that I've seen where Node is used this way, they are definitely conflated. They present some attributes for listener selection in Node metadata, but these attributes are irrelevant elsewhere. 

I'm also opposed to the overloading of Node for cacheability. If we want to do things like xDS relay, but the management server is conditioning the resource returned based on Node, this effectively drags the entire Node message in to cache key. This seems strange; it would be like including the IP address of a specific client web browser in the cache key for a CDN.

Reply:
In some sense, Node is a namespace that is "global" to all xDS requests coming from the client. The substantial downside of having a namespace per xDS config source is that in some deployments the management server will have to echo back the passed namespace in an xDS configured config source (e.g. LDS returning RDS). This is not ergonomic or easily possible in some cases. Because of this, I think you are always going to need some type of "default namespace" or "global namespace" that is configurable in the bootstrap. This can be overridden/added to by individual config sources, but it would be passed with all xDS requests originating from the client. So, yes, we can add this directly to bootstrap and be explicit about it, but it's really no different than Node is today.

So I'm in favor of cleaning up the logical discrepancy between Node/global namespace, but I feel very strongly that we cannot get rid of it as a concept.

Reply:
I'd like to drill down on the "ergonomic or easily possible in some cases.". This seems to be the crux of the discussion.

Reply:
> I'd like to drill down on the "ergonomic or easily possible in some cases.". This seems to be the crux of the discussion.

If a deployment is keying the xDS resources at least partially based on hierarchical node/namespace information (e.g. something like `canary.foo_service.bar_zone.baz_region` in reverse DNS format), and we implicitly remove the ability to carry this along with each xDS request, the management server now has to become more complicated and more aware of the client topology than it otherwise would be. Concretely, in a simple deployment, it would have to inject this namespace into child config sources that are returned as part of previous xDS responses. Clearly this can be done, but it has caching and complexity implications that seem silly. IMO a (partial) global namespace defined in bootstrap is a very natural way of expressing intention and keeping the management server simple. I very much like the idea of per config_source namespace that can override/add to the (optional) global namespace, but as I said it makes very little sense to me to remove this conceptual functionality.